File: native/apple/Sources/CoreStorage/CloudSync.swift
import Foundation

/// Conflict winner when local and remote entries differ for the same identifier.
public enum HistorySyncConflictResolution: String, Sendable, Hashable, Codable {
    case local
    case remote
}

/// Conflict metadata for one merged history identifier.
public struct HistorySyncConflict: Sendable, Hashable, Codable {
    public let id: String
    public let resolution: HistorySyncConflictResolution
    public let localCreatedAtMillis: Int64
    public let remoteCreatedAtMillis: Int64

    public init(
        id: String,
        resolution: HistorySyncConflictResolution,
        localCreatedAtMillis: Int64,
        remoteCreatedAtMillis: Int64
    ) {
        self.id = id
        self.resolution = resolution
        self.localCreatedAtMillis = localCreatedAtMillis
        self.remoteCreatedAtMillis = remoteCreatedAtMillis
    }
}

/// Aggregate sync statistics for one cloud sync operation.
public struct HistorySyncStats: Sendable, Hashable, Codable {
    public var added: Int
    public var updated: Int
    public var unchanged: Int
    public var conflicts: Int

    public init(added: Int, updated: Int, unchanged: Int, conflicts: Int) {
        self.added = added
        self.updated = updated
        self.unchanged = unchanged
        self.conflicts = conflicts
    }
}

/// Result payload returned after a sync operation completes.
public struct HistorySyncResult: Sendable, Hashable, Codable {
    public let stats: HistorySyncStats
    public let conflicts: [HistorySyncConflict]
    public let syncedAtMillis: Int64

    public init(
        stats: HistorySyncStats,
        conflicts: [HistorySyncConflict],
        syncedAtMillis: Int64
    ) {
        self.stats = stats
        self.conflicts = conflicts
        self.syncedAtMillis = syncedAtMillis
    }
}

/// Cloud adapter contract for loading and persisting history snapshots.
public protocol HistoryCloudSyncAdapter: Sendable {
    func fetchRemoteEntries() async throws -> [HistoryEntry]
    func pushRemoteEntries(_ entries: [HistoryEntry]) async throws
}

/// Errors raised while reading or writing sync payloads.
public enum HistoryCloudSyncError: LocalizedError {
    case invalidPayloadVersion(Int)
    case payloadDecodeFailed

    public var errorDescription: String? {
        switch self {
        case let .invalidPayloadVersion(version):
            return "Unsupported cloud sync payload version: \(version)."
        case .payloadDecodeFailed:
            return "Unable to decode cloud sync payload."
        }
    }
}

/// iCloud-backed history adapter using ubiquitous key-value storage.
public actor ICloudHistorySyncAdapter: HistoryCloudSyncAdapter {
    private struct CloudHistoryPayload: Codable {
        let version: Int
        let exportedAtMillis: Int64
        let entries: [HistoryEntry]
    }

    private let store: NSUbiquitousKeyValueStore
    private let storageKey: String
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private let nowMillis: @Sendable () -> Int64

    public init(
        store: NSUbiquitousKeyValueStore = .default,
        storageKey: String = "secure_pastebin_history_cloud_payload_v1",
        nowMillis: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.store = store
        self.storageKey = storageKey
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
        self.nowMillis = nowMillis
    }

    public func fetchRemoteEntries() async throws -> [HistoryEntry] {
        guard let data = store.data(forKey: storageKey) else {
            return []
        }

        guard let payload = try? decoder.decode(CloudHistoryPayload.self, from: data) else {
            throw HistoryCloudSyncError.payloadDecodeFailed
        }
        guard payload.version == 1 else {
            throw HistoryCloudSyncError.invalidPayloadVersion(payload.version)
        }
        return payload.entries
    }

    public func pushRemoteEntries(_ entries: [HistoryEntry]) async throws {
        let payload = CloudHistoryPayload(
            version: 1,
            exportedAtMillis: nowMillis(),
            entries: entries
        )
        let data = try encoder.encode(payload)
        store.set(data, forKey: storageKey)
        _ = store.synchronize()
    }
}

/// Coordinates one-shot history sync and conflict-aware merge.
public actor HistoryCloudSyncCoordinator {
    private let historyStore: HistoryStore
    private let cloudAdapter: HistoryCloudSyncAdapter
    private let nowMillis: @Sendable () -> Int64

    public init(
        historyStore: HistoryStore,
        cloudAdapter: HistoryCloudSyncAdapter,
        nowMillis: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.historyStore = historyStore
        self.cloudAdapter = cloudAdapter
        self.nowMillis = nowMillis
    }

    /// Syncs local and cloud history snapshots and returns merge/conflict metadata.
    public func syncNow() async throws -> HistorySyncResult {
        let localEntries = try await historyStore.list()
        let remoteEntries = try await cloudAdapter.fetchRemoteEntries()
        let merge = mergeEntries(local: localEntries, remote: remoteEntries)

        let localIDs = Set(localEntries.map(\.id))
        let mergedIDs = Set(merge.entries.map(\.id))
        for removedID in localIDs.subtracting(mergedIDs) {
            try await historyStore.delete(id: removedID)
        }
        for entry in merge.entries {
            try await historyStore.upsert(entry)
        }
        try await cloudAdapter.pushRemoteEntries(merge.entries)

        return HistorySyncResult(
            stats: merge.stats,
            conflicts: merge.conflicts,
            syncedAtMillis: nowMillis()
        )
    }

    private func mergeEntries(
        local: [HistoryEntry],
        remote: [HistoryEntry]
    ) -> (entries: [HistoryEntry], stats: HistorySyncStats, conflicts: [HistorySyncConflict]) {
        let localMap = Dictionary(uniqueKeysWithValues: local.map { ($0.id, $0) })
        let remoteMap = Dictionary(uniqueKeysWithValues: remote.map { ($0.id, $0) })
        let allIDs = Set(localMap.keys).union(remoteMap.keys)

        var merged: [HistoryEntry] = []
        var conflicts: [HistorySyncConflict] = []
        var stats = HistorySyncStats(added: 0, updated: 0, unchanged: 0, conflicts: 0)

        for id in allIDs {
            let localEntry = localMap[id]
            let remoteEntry = remoteMap[id]

            switch (localEntry, remoteEntry) {
            case let (localEntry?, nil):
                merged.append(localEntry)
                stats.added += 1
            case let (nil, remoteEntry?):
                merged.append(remoteEntry)
                stats.added += 1
            case let (localEntry?, remoteEntry?):
                if localEntry == remoteEntry {
                    merged.append(localEntry)
                    stats.unchanged += 1
                    continue
                }

                let resolution: HistorySyncConflictResolution
                let winner: HistoryEntry
                if localEntry.createdAtMillis >= remoteEntry.createdAtMillis {
                    resolution = .local
                    winner = localEntry
                } else {
                    resolution = .remote
                    winner = remoteEntry
                }

                merged.append(winner)
                conflicts.append(
                    HistorySyncConflict(
                        id: id,
                        resolution: resolution,
                        localCreatedAtMillis: localEntry.createdAtMillis,
                        remoteCreatedAtMillis: remoteEntry.createdAtMillis
                    )
                )
                stats.updated += 1
                stats.conflicts += 1
            case (nil, nil):
                continue
            }
        }

        merged.sort { $0.createdAtMillis > $1.createdAtMillis }
        return (entries: merged, stats: stats, conflicts: conflicts)
    }
}

File: native/android/core/storage/src/main/kotlin/com/securepastebin/core/storage/CloudSync.kt
package com.securepastebin.core.storage

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import org.json.JSONArray
import org.json.JSONObject

/**
 * Conflict winner when local and remote entries differ for one ID.
 */
enum class HistorySyncConflictResolution {
    LOCAL,
    REMOTE,
}

/**
 * One conflict resolution record for cloud merge operations.
 */
data class HistorySyncConflict(
    val id: String,
    val resolution: HistorySyncConflictResolution,
    val localCreatedAtMillis: Long,
    val remoteCreatedAtMillis: Long,
)

/**
 * Aggregate sync counters for one sync execution.
 */
data class HistorySyncStats(
    val added: Int,
    val updated: Int,
    val unchanged: Int,
    val conflicts: Int,
)

/**
 * Cloud sync result returned by the coordinator.
 */
data class HistorySyncResult(
    val stats: HistorySyncStats,
    val conflicts: List<HistorySyncConflict>,
    val syncedAtMillis: Long,
)

/**
 * Cloud adapter contract for loading and persisting history snapshots.
 */
interface HistoryCloudSyncAdapter {
    suspend fun fetchRemoteEntries(): List<HistoryEntry>
    suspend fun pushRemoteEntries(entries: List<HistoryEntry>)
}

/**
 * Google Drive document-based cloud adapter using Storage Access Framework URIs.
 */
class GoogleDriveHistorySyncAdapter(
    private val context: Context,
    private val documentUri: Uri,
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) : HistoryCloudSyncAdapter {
    override suspend fun fetchRemoteEntries(): List<HistoryEntry> {
        val raw = context.contentResolver.openInputStream(documentUri)?.use { input ->
            input.readBytes().toString(Charsets.UTF_8)
        } ?: return emptyList()

        if (raw.isBlank()) {
            return emptyList()
        }

        return parseEntries(raw)
    }

    override suspend fun pushRemoteEntries(entries: List<HistoryEntry>) {
        val payload = JSONObject().apply {
            put("version", 1)
            put("exportedAtMillis", nowMillis())
            put("entries", entriesToJSONArray(entries))
        }

        context.contentResolver.openOutputStream(documentUri, "wt")?.use { output ->
            output.write(payload.toString().toByteArray(Charsets.UTF_8))
        } ?: throw IllegalStateException("Unable to open Google Drive sync file for writing.")
    }

    private fun parseEntries(raw: String): List<HistoryEntry> {
        return runCatching {
            val trimmed = raw.trim()
            if (trimmed.startsWith("[")) {
                parseEntriesArray(JSONArray(trimmed))
            } else {
                val payload = JSONObject(trimmed)
                val version = payload.optInt("version", 0)
                require(version == 1) { "Unsupported cloud sync payload version: $version" }
                parseEntriesArray(payload.optJSONArray("entries") ?: JSONArray())
            }
        }.getOrElse { throwable ->
            throw IllegalStateException("Unable to parse Google Drive sync payload.", throwable)
        }
    }

    private fun parseEntriesArray(array: JSONArray): List<HistoryEntry> {
        return buildList {
            for (index in 0 until array.length()) {
                val item = array.optJSONObject(index) ?: continue
                val id = item.optString("id")
                val fileName = item.optString("fileName")
                val createdAtMillis = item.optLong("createdAtMillis")
                val expiresAtMillis = item.optLong("expiresAtMillis")
                if (id.isBlank() || fileName.isBlank()) {
                    continue
                }
                add(
                    HistoryEntry(
                        id = id,
                        fileName = fileName,
                        createdAtMillis = createdAtMillis,
                        expiresAtMillis = expiresAtMillis,
                    ),
                )
            }
        }
    }

    private fun entriesToJSONArray(entries: List<HistoryEntry>): JSONArray {
        val array = JSONArray()
        entries.forEach { entry ->
            array.put(
                JSONObject().apply {
                    put("id", entry.id)
                    put("fileName", entry.fileName)
                    put("createdAtMillis", entry.createdAtMillis)
                    put("expiresAtMillis", entry.expiresAtMillis)
                },
            )
        }
        return array
    }
}

/**
 * One-shot history cloud sync coordinator with conflict-aware merge tracking.
 */
class HistoryCloudSyncCoordinator(
    private val historyStore: HistoryStore,
    private val cloudAdapter: HistoryCloudSyncAdapter,
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) {
    suspend fun syncNow(): HistorySyncResult {
        val localEntries = historyStore.list()
        val remoteEntries = cloudAdapter.fetchRemoteEntries()
        val merge = mergeEntries(local = localEntries, remote = remoteEntries)

        val localIDs = localEntries.map { it.id }.toSet()
        val mergedIDs = merge.entries.map { it.id }.toSet()
        localIDs.subtract(mergedIDs).forEach { removedID ->
            historyStore.delete(removedID)
        }
        merge.entries.forEach { historyStore.upsert(it) }
        cloudAdapter.pushRemoteEntries(merge.entries)

        return HistorySyncResult(
            stats = merge.stats,
            conflicts = merge.conflicts,
            syncedAtMillis = nowMillis(),
        )
    }

    private data class MergeAccumulator(
        val entries: List<HistoryEntry>,
        val stats: HistorySyncStats,
        val conflicts: List<HistorySyncConflict>,
    )

    private fun mergeEntries(
        local: List<HistoryEntry>,
        remote: List<HistoryEntry>,
    ): MergeAccumulator {
        val localMap = local.associateBy { it.id }
        val remoteMap = remote.associateBy { it.id }
        val allIDs = localMap.keys + remoteMap.keys

        val merged = mutableListOf<HistoryEntry>()
        val conflicts = mutableListOf<HistorySyncConflict>()
        var added = 0
        var updated = 0
        var unchanged = 0
        var conflictCount = 0

        allIDs.forEach { id ->
            val localEntry = localMap[id]
            val remoteEntry = remoteMap[id]

            when {
                localEntry != null && remoteEntry == null -> {
                    merged += localEntry
                    added += 1
                }
                localEntry == null && remoteEntry != null -> {
                    merged += remoteEntry
                    added += 1
                }
                localEntry != null && remoteEntry != null -> {
                    if (localEntry == remoteEntry) {
                        merged += localEntry
                        unchanged += 1
                    } else {
                        val resolution: HistorySyncConflictResolution
                        val winner: HistoryEntry
                        if (localEntry.createdAtMillis >= remoteEntry.createdAtMillis) {
                            resolution = HistorySyncConflictResolution.LOCAL
                            winner = localEntry
                        } else {
                            resolution = HistorySyncConflictResolution.REMOTE
                            winner = remoteEntry
                        }

                        merged += winner
                        conflicts += HistorySyncConflict(
                            id = id,
                            resolution = resolution,
                            localCreatedAtMillis = localEntry.createdAtMillis,
                            remoteCreatedAtMillis = remoteEntry.createdAtMillis,
                        )
                        updated += 1
                        conflictCount += 1
                    }
                }
            }
        }

        return MergeAccumulator(
            entries = merged.sortedByDescending { it.createdAtMillis },
            stats = HistorySyncStats(
                added = added,
                updated = updated,
                unchanged = unchanged,
                conflicts = conflictCount,
            ),
            conflicts = conflicts,
        )
    }
}

/**
 * Utility to persist and restore cloud sync URI selections.
 */
class GoogleDriveSyncConfigurationStore(
    context: Context,
    private val preferenceName: String = "secure_pastebin_cloud_sync_v1",
    private val documentURIKey: String = "google_drive_sync_document_uri",
) {
    private val sharedPreferences =
        context.getSharedPreferences(preferenceName, Context.MODE_PRIVATE)

    fun readDocumentURI(): String? {
        return sharedPreferences.getString(documentURIKey, null)
    }

    fun writeDocumentURI(uri: Uri) {
        sharedPreferences.edit().putString(documentURIKey, uri.toString()).apply()
    }

    fun clear() {
        sharedPreferences.edit().remove(documentURIKey).apply()
    }

    fun takePersistablePermissions(
        contentResolver: ContentResolver,
        uri: Uri,
    ) {
        runCatching {
            contentResolver.takePersistableUriPermission(
                uri,
                android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION or
                    android.content.Intent.FLAG_GRANT_WRITE_URI_PERMISSION,
            )
        }
    }
}
